### ISSUE TITLE
UDP Sockets

### ISSUE THREAD

**Issue by softworkz** [mgravell/Pipelines.Sockets.Unofficial#36]
One question: 

Is this supposed to work with UDP sockets as well? 

Did somebody test this?

Thanks,

softworkz

**Comment by mgravell** [comment]
UDP sockets aren't streams in the same way and would presumably need extra
consideration. Feel free to test it!

**Comment by softworkz** [comment]
I did.

To make it work in general it requires just a single line change:

https://github.com/mgravell/Pipelines.Sockets.Unofficial/blob/ea683bfc007ee4adba351d42a8438af5239225c1/src/Pipelines.Sockets.Unofficial/SocketConnection.Receive.cs#L54

Needs to be changed to this:

```
var buffer = _receiveFromSocket.Writer.GetMemory(this._receiveOptions.MinimumSegmentSize);
```

And ZeroLength reads must not be used.

But overall it didn't perform very well due to the frequent flushing (= after each receive).

With UDP, the data doesn't get accumulated in the buffer. You get a single datagram with each receive call. And that causes too many async calls which are becoming a performance issue with high data rates.

**Comment by Drawaes** [comment]
I'll be honest as a big user of UDP and a user of pipelines I had considered it. However as Marc said it doesn't make a lot of sense in this case, you completely lose the framing of the datagram so then you need to implement framing over the top, at which point I question if you are rewriting tcp over UDP.

**Comment by softworkz** [comment]
No, the framing doesn't get lost as mentioned above because in case of UDP you always get a single datagram with each receive call. A datagram contains n packets of size x with x being a global constant. Packets are not sliced across datagrams.

Maybe I should say that this is about media streaming. Several protocols in this area are UDP based and the challenge is to find the most effective implementation.

Arguing about protocol design and choice is a dead end in this case.. :-)

**Comment by mgravell** [comment]
To me, UDP seems more like a `Channel<ReadOnlySequence<byte>>`. Perhaps that's something that we could implement with minimal work, giving you all the benefits without the pain, and respecting the framing differences between UDP and TCP.

Happy to explore that area. I've mostly been TCP focused, but I love expanding my horizons...

**Comment by Drawaes** [comment]
Which was exactly the discussion I had with Fowler around UDP. I suspect it's not on the pipelines asp.net radar yet either (maybe quic will change that)

**Comment by softworkz** [comment]
What do you mean by Channel<ReadOnlySequence<byte>> ?
Wasn't 'Channel' the initial term before it became 'Pipeline'?

**Comment by Drawaes** [comment]
System.Threading.Channels made around the same time as pipelines (which maybe why pipelines was renamed... But that's just a rumour 😀)

**Comment by mgravell** [comment]
`Channel<T>` is the high perf reader-writer async queue thing. Yes, the name was also the early name of pipelines, but pipelines lost and became pipelines. Related and overlapping topic, and I do think it would make a good API here, for packet-based rather than frame-based payloads.

**Comment by softworkz** [comment]
That sounds really interesting, every time I read about channels somewhere I thought it was about the early pipeline stuff.

Too bad that it doesn't seem to support netstandard2.0 (need to support BSD as well)

**Comment by mgravell** [comment]
Channels and pipelines are both netstandard 1.3+; you're fine

**Comment by softworkz** [comment]
So, it's not yet documented as such?

Or is that the wrong place: https://docs.microsoft.com/en-us/dotnet/api/system.threading.channels.channel-1?view=dotnet-plat-ext-2.1&viewFallbackFrom=netstandard-2.0

**Comment by mgravell** [comment]
Nuget: https://www.nuget.org/packages/System.Threading.Channels/ - click dependencies

**Comment by softworkz** [comment]
Thanks a lot, I'll look into it!

**Comment by mgravell** [comment]
In fact, maybe a case could be made for `IAsyncEnumerable<ReadOnlySequence<byte>>` - I will have to play!

**Comment by softworkz** [comment]
Think about small packet sizes and a huge amount of them :-)

**Comment by softworkz** [comment]
Don't forget to add a running number when sending udp datagrams for testing.
You need to make sure that you're getting them all. 

Mine are about 25k size, I tried to get close to half of the maximum allowed (64k).
The smaller I made the datagrams the worse it got (CPU usage).
(it's a real world application - that means: half datagram size => double amount of datagrams per time)

PS: Watch your e-mail, I can provide some more interesting details if you wish

**Comment by mgravell** [comment]
Started a spike with this. Looking viable. A running number in the lib would be pointless - it can't tell if it didn't get a packet. Hence IMO any running number *must* be part of your payload, not the code that makes things work.

**Comment by softworkz** [comment]
Yes, of course I meant a number inside the payload for testing.

**Comment by sebandraos** [comment]
Did anything come of this discussion? I can see a UDP branch but it doesn't seem like it ever got merged into master. Was an alternative to that branch implemented instead? Thanks

**Comment by mgravell** [comment]
At the current time: no. It is of interest, but I have 72 other things that
I'm working on, and it isn't at the top of my list. I'm so aware that
Bedrock is (hopefully) gaining some client-side love for .NET 5, so I was
hoping to see if anything changes there.

**Comment by sebandraos** [comment]
Ok, cool thanks.

**Comment by sgf** [comment]
The TCP protocol performs very badly in the case of a wide geographical distance around the world. 
UDP is controlled by appropriate high-level algorithms to achieve an ideal low latency.
